bapm_server/bapm_publisher.py

```diff
@@ -0,0 +1,246 @@
+        )
+
+        message = {"test msg": "test"}
+        self._channel.basic_publish(
+            self.EXCHANGE,
+            self.ROUTING_KEY,
+            json.dumps(message, ensure_ascii=False),
+            properties,
+        )
+        self._message_number += 1
+        self._deliveries.append(self._message_number)
+        LOGGER.info("Published message # %i", self._message_number)
+        self.schedule_next_message()
+
+    def run(self):
+        """Run the example code by connecting and then starting the IOLoop."""
+        self._connection = None
+        self._deliveries = []
+        self._acked = 0
+        self._nacked = 0
+        self._message_number = 0
+
+        try:
+            self._connection = self.connect()
+            self._connection.ioloop.start()
+        except KeyboardInterrupt:
+            self.stop()
+            if self._connection is not None and not self._connection.is_closed:
+                # Finish closing
+                self._connection.ioloop.start()
+
+        # while not self._stopping:
+        #     self._connection = None
+        #     self._deliveries = []
+        #     self._acked = 0
+        #     self._nacked = 0
+        #     self._message_number = 0
+
+        #     try:
+        #         self._connection = self.connect()
+        #         self._connection.ioloop.start()
+        #     except KeyboardInterrupt:
+        #         self.stop()
+        #         if self._connection is not None and not self._connection.is_closed:
+        #             # Finish closing
+        #             self._connection.ioloop.start()
+
+        LOGGER.info("Stopped")
+
+    def stop(self):
+        """Stop the example by closing the channel and connection. We
+        set a flag here so that we stop scheduling new messages to be
+        published. The IOLoop is started because this method is
+        invoked by the Try/Catch below when KeyboardInterrupt is caught.
+        Starting the IOLoop again will allow the publisher to cleanly
+        disconnect from RabbitMQ.
+
+        """
+        LOGGER.info("Stopping")
+        self._stopping = True
+        self.close_channel()
+        self.close_connection()
+
+    def close_channel(self):
+        """Invoke this command to close the channel with RabbitMQ by sending
+        the Channel.Close RPC command.
+
+        """
+        if self._channel is not None:
+            LOGGER.info("Closing the channel")
+            self._channel.close()
+
+    def close_connection(self):
+        """This method closes the connection to RabbitMQ."""
+        if self._connection is not None:
+            LOGGER.info("Closing connection")
+            self._connection.close()
+
+
+def main():
+    logging.basicConfig(level=logging.DEBUG, format=LOG_FORMAT)
+
+    # Connect to localhost:5672 as guest with the password guest and virtual host "/" (%2F)
+    publisher = BAPMPublisher("amqp://guest:guest@aw-sdx-monitor.renci.org:5672/%2F")
```

Comment:
We're leaking secrets here too.  If it is just for testing/example, it probably is fine, but we should avoid such hard-coding.

Reply:
Yes we will add RabbitMQ security check, but for now it should be fine. The main() is for testing purposes only.